src/search/algorithms/named_vector.h

```diff
@@ -0,0 +1,103 @@
+#ifndef ALGORITHMS_NAMED_VECTOR_H
+#define ALGORITHMS_NAMED_VECTOR_H
+
+#include <string>
+#include <vector>
+
+namespace named_vector {
+/*
+  NamedVector is a vector-like collection with optional names to be associated with each element.
+  It is optimized to have minimal overhead when there are no names.
+  Any name which is not specified is assumed to be an empty string.
+ */
+template<typename T>
+class NamedVector {
+    std::vector<T> elements;
+    std::vector<std::string> names;
+public:
+    NamedVector() = default;
+    NamedVector(NamedVector<T> &&other) = default;
+    NamedVector(NamedVector<T> &other) = default;
+
+    template<typename ... _Args>
+    void emplace_back(_Args && ... __args) {
+        elements.emplace_back(std::forward<_Args>(__args) ...);
+    }
+
+    void push_back(const T &element) {
+        elements.push_back(element);
+    }
+
+    T &operator[](int index) {
+        return elements[index];
+    }
+
+    bool has_names() const {
+        return !names.empty();
+    }
+
+    void set_name(int index, const std::string &name) {
```

Comment:
Getting and setting a name for an invalid index should be an error. You can use `utils::in_bounds` to check the index and `utils::ExitCodes::SearchCriticalError` (I think) for exiting. Look for other places where exit codes from `utils::ExitCodes` are used to see the usage.

Reply:
I would assert rather than test and exit. This is more consistent with other parts of the code and avoids an unnecessary dependency of the algorithms (which are not meant to be specific to Fast Downward) on Fast Downward's exit code.